core/src/container.cpp

```diff
@@ -197,43 +197,56 @@ void ContainerBasePrivate::copyState(Interface::iterator external, const Interfa
 
 	// create a clone of external state within target interface (child's starts() or ends())
 	auto internal = states_.insert(states_.end(), InterfaceState(*external));
-	target->add(*internal);
 	// and remember the mapping between them
 	internalToExternalMap().insert(std::make_pair(&*internal, &*external));
+	target->add(*internal);
 }
 
 void ContainerBasePrivate::liftSolution(const SolutionBasePtr& solution, const InterfaceState* internal_from,
-                                        const InterfaceState* internal_to) {
-	computeCost(*internal_from, *internal_to, *solution);
-
-	// map internal to external states
-	auto find_or_create_external = [this](const InterfaceState* internal, bool& created) -> InterfaceState* {
-		auto it = internalToExternalMap().find(internal);
-		if (it != internalToExternalMap().end())
-			return const_cast<InterfaceState*>(it->second);
-
-		InterfaceState* external = &*states_.insert(states_.end(), InterfaceState(*internal));
-		internalToExternalMap().insert(std::make_pair(internal, external));
-		created = true;
+                                        const InterfaceState* internal_to, InterfaceState* new_from,
+                                        InterfaceState* new_to) {
+	const bool create_from{ requiredInterface().testFlag(WRITES_PREV_END) };
+	const bool create_to{ requiredInterface().testFlag(WRITES_NEXT_START) };
+
+	// external states, nullptr if they don't exist yet
+	const InterfaceState* external_from{ create_from ? new_from : internalToExternalMap().at(internal_from) };
+	const InterfaceState* external_to{ create_to ? new_to : internalToExternalMap().at(internal_to) };
+
+	// computeCost
+	// we can pass intern_{from/to} here because in this case the lifted states that might be created later
+	// are equivalent up to the connected Solutions (which are not relevant for CostTerms)
+	computeCost(external_from ? *external_from : *internal_from, external_to ? *external_to : *internal_to, *solution);
+
+	// storeSolution
+	if (!storeSolution(solution, external_from, external_to)) {
+		return;
+	}
+
+	auto create_state = [this](const InterfaceState& internal, InterfaceState* new_external) {
+		InterfaceState* external{ storeState(new_external ? std::move(*new_external) : InterfaceState{ internal }) };
+		internalToExternalMap().insert(std::make_pair(&internal, external));
```

Comment:
If `new_external` is `nullptr`, a new copy of `internal` will be created unconditionally here, no matter we already encountered this state before or not.
If it's non-null, you don't check whether it's the same one encountered before for this `internal` state.